Formula/glibc.rb

```diff
@@ -95,41 +95,188 @@ class Glibc < Formula
   depends_on BrewedGlibcNotOlderRequirement
   depends_on :linux
   depends_on LinuxKernelRequirement
 
-  # GCC 4.7 or later is required.
-  fails_with gcc: "4.6"
+  MINIMUM_MAKE_VERSION = "4.0".freeze
+  MINIMUM_GCC_VERSION = "6.2".freeze
+  INSTALLED_MAKE_VERSION = Utils.safe_popen_read("make", "--version").lines.first.split.third.freeze
+  INSTALLED_GCC_VERSION = Utils.safe_popen_read("gcc", "-dumpversion").freeze
 
-  def install
-    # Fix Error: `loc1@GLIBC_2.2.5' can't be versioned to common symbol 'loc1'
-    # See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=869717
-    # Fixed in glibc 2.24
-    inreplace "misc/regexp.c", /^(char \*loc[12s]);$/, "\\1 __attribute__ ((nocommon));"
+  if Version.new(INSTALLED_MAKE_VERSION) < MINIMUM_MAKE_VERSION
+    resource "make" do
+      url "https://ftp.gnu.org/gnu/make/make-4.3.tar.gz"
+      sha256 "e05fdde47c5f7ca45cb697e973894ff4f5d79e13b750ed57d7b66d8defc78e19"
+    end
+  end
+
+  if which("python3").blank?
+    resource "python3" do
+      url "https://www.python.org/ftp/python/3.10.5/Python-3.10.5.tgz"
+      sha256 "18f57182a2de3b0be76dfc39fdcfd28156bb6dd23e5f08696f7492e9e3d0bf2d"
+    end
+  end
+
+  if which("bison").blank?
+    if which("m4").blank?
+      resource "m4" do
+        url "https://ftp.gnu.org/gnu/m4/m4-1.4.19.tar.xz"
+        sha256 "63aede5c6d33b6d9b13511cd0be2cac046f2e70fd0a07aa9573a04a82783af96"
+      end
+    end
+
+    resource "bison" do
+      url "https://ftp.gnu.org/gnu/bison/bison-3.8.2.tar.xz"
+      sha256 "9bba0214ccf7f1079c5d59210045227bcf619519840ebfa80cd3849cff5a5bf2"
+    end
+  end
+
+  if Version.new(INSTALLED_GCC_VERSION) < MINIMUM_GCC_VERSION
+    unless File.exist?("/usr/bin/bzip2")
+      resource "bzip2" do
+        url "https://sourceware.org/pub/bzip2/bzip2-1.0.8.tar.gz"
+        sha256 "ab5a03176ee106d3f0fa90e381da478ddae405918153cca248e682cd0c4a2269"
+      end
+    end
+
+    resource "gcc" do
+      url "https://ftp.gnu.org/gnu/gcc/gcc-10.4.0/gcc-10.4.0.tar.xz"
+      sha256 "c9297d5bcd7cb43f3dfc2fed5389e948c9312fd962ef6a4ce455cff963ebe4f1"
+    end
+
+    resource "mpfr" do
+      url "ftp://gcc.gnu.org/pub/gcc/infrastructure/mpfr-3.1.6.tar.bz2"
+      sha256 "cf4f4b2d80abb79e820e78c8077b6725bbbb4e8f41896783c899087be0e94068"
+    end
+
+    resource "mpc" do
+      url "ftp://gcc.gnu.org/pub/gcc/infrastructure/mpc-1.0.3.tar.gz"
+      sha256 "617decc6ea09889fb08ede330917a00b16809b8db88c29c31bfbb49cbf88ecc3"
+    end
+
+    resource "gmp" do
+      url "ftp://gcc.gnu.org/pub/gcc/infrastructure/gmp-6.1.0.tar.bz2"
+      sha256 "498449a994efeba527885c10405993427995d3f86b8768d8cdf8d9dd7c6b73e8"
+    end
+  end
 
+  def install
     # Setting RPATH breaks glibc.
     %w[
       LDFLAGS LD_LIBRARY_PATH LD_RUN_PATH LIBRARY_PATH
       HOMEBREW_DYNAMIC_LINKER HOMEBREW_LIBRARY_PATHS HOMEBREW_RPATH_PATHS
     ].each { |x| ENV.delete x }
 
+    ENV["CC"] = DevelopmentTools.locate("gcc")
+    ENV["CXX"] = DevelopmentTools.locate("g++")
```

Comment:
Isn't this the default behaviour?

Reply:
I think the important thing is that it has to be `gcc` and not `gcc-5` etc.  because those could brewed GCC which won’t work.  Although I’m now wondering if that could be addressed in a Requirement by dumping the GCC specs and scanning them for `rpath` and rejecting that compiler if `rpath` is found, even if it is new enough.  That would provide maximum flexibility in choice of compiler.  But I need to see how all of this comes together.

Reply:
> because those could brewed GCC which won’t work

Sorry, don't understand this bit! Can you elaborate?

> Although I’m now wondering if that could be addressed in a Requirement by dumping the GCC specs and scanning them for rpath and rejecting that compiler if rpath is found, even if it is new enough.

Yeh, that seems ideal.

Reply:
Brewed GCC doesn't work because we add an RPATH into our custom spec, but the `ld.so` binary in `glibc` will break if an RPATH is added because it *is* the dynamic linker and cannot use itself to find run time libraries. `glibc` has no external run time dependencies and uses a different approach for its libraries to find each other internally.  

We also cannot use brewed GCC to build the vendored dependencies.  Although they do allow RPATHs in general, one of the RPATHs we add is to the `opt_lib` of brewed `glibc`, which cause the build dependencies to start mixing in parts of `glibc` as it is building which eventually breaks them part way through the build.

If we just use vendored dependencies unconditionally as I would prefer this isn't really an issue.  `DevelopmentTools.locate("gcc")` will never return a brewed GCC because all of our brewed GCC end in version suffixes like `gcc-5`, `g++-5`, etc.  This is exactly what we want so that when users upgrade from version of `glibc` to another, any brewed GCC they have installed will be ignored.  This concern is obviously irrelevant to new installations but does matter for existing ones.

Reply:
> Brewed GCC doesn't work because we add an RPATH into our custom spec

Can this be changed at runtime? If not, could we consider changing this in `brew test-bot` or something? Seems a lot to rebuild GCC just because of a custom spec.

> If we just use vendored dependencies unconditionally as I would prefer this isn't really an issue.

What about system dependencies unconditionally? Would that solve any of this?

Reply:
Just to be clear, the issue with RPATHs from brewed GCC is specifically a concern for existing installations.  If the user tried to use brewed GCC to build the `glibc` build requirements outside of the formula, they would still break the because brewed GCC uses its custom spec at all times, even outside of `brew`, and in all other cases this is a good thing.

I am going to see if it's possible to strip out the `rpath` arguments added by the spec using the superenv when the formula name includes `glibc`.  If that works we don't have to anything special at all in the formula - we just unconditionally vendor GCC and don't have to worry about whatever GCC brew initially picks breaking anything.